File: Vialer/APNS/APNSHandler.m
//
//  APNSHandler.m
//  Copyright © 2016 voipgrid.com. All rights reserved.
//

#import "APNSHandler.h"
#import "Middleware.h"
#import "SystemUser.h"

@interface APNSHandler ()
@property (strong, nonatomic) PKPushRegistry *voipRegistry;
@property (strong, nonatomic) Middleware *middleware;
@end

@implementation APNSHandler

// To make the singleton pattern testable.
static APNSHandler *_sharedAPNSHandler = nil;
static dispatch_once_t onceToken = 0;

#pragma mark - Lifecycle
+ (instancetype)sharedHandler {
    dispatch_once(&onceToken, ^{
        _sharedAPNSHandler = [[self alloc] init];
    });
    return _sharedAPNSHandler;
}

+ (void)setSharedHandler:(APNSHandler *)sharedHandler {
    if (_sharedAPNSHandler != sharedHandler) {
        _sharedAPNSHandler = sharedHandler;

        if (!_sharedAPNSHandler) {
            onceToken = 0;
        } else {
            onceToken = -1;
        }
    }
}

#pragma mark - properties
- (PKPushRegistry *)voipRegistry {
    if (!_voipRegistry) {
        _voipRegistry = [[PKPushRegistry alloc] initWithQueue:nil];
    }
    return _voipRegistry;
}

- (Middleware *)middleware {
    if (!_middleware) {
        _middleware = [[Middleware alloc] init];
    }
    return _middleware;
}


#pragma mark - actions
- (void)registerForVoIPNotifications {
    // Only register once, if delegate is set, registration has been done before:
    if (!self.voipRegistry.delegate) {
        self.voipRegistry.delegate = self;

        VialerLogVerbose(@"Initiating VoIP push registration");
        self.voipRegistry.desiredPushTypes = [NSSet setWithObject:PKPushTypeVoIP];
    }
}

+ (NSString *)storedAPNSToken {
    APNSHandler *sharedHandler = [[self class] sharedHandler];
    NSData *token = [sharedHandler.voipRegistry pushTokenForType:PKPushTypeVoIP];
    return [sharedHandler nsStringFromNSData:token];
}

#pragma mark - PKPushRegistray management
- (void)pushRegistry:(PKPushRegistry *)registry didInvalidatePushTokenForType:(NSString *)type {
    VialerLogWarning(@"APNS Token became invalid");
}

- (void)pushRegistry:(PKPushRegistry *)registry didReceiveIncomingPushWithPayload:(PKPushPayload *)payload forType:(NSString *)type {
    VialerLogDebug(@"Incoming push notification of type: %@", type);
    [self.middleware handleReceivedAPSNPayload:[payload dictionaryPayload]];
}

- (void)pushRegistry:(PKPushRegistry *)registry didUpdatePushCredentials:(PKPushCredentials *)credentials forType:(NSString *)type {
    VialerLogInfo(@"Type:%@. APNS registration successful. Token: %@", type, credentials.token);
    [self.middleware sentAPNSToken:[self nsStringFromNSData:credentials.token]];
}

#pragma mark - token conversion
/*
 * Returns hexadecimal string of NSData. Empty string if data is empty.
 * http://stackoverflow.com/questions/1305225/best-way-to-serialize-a-nsdata-into-an-hexadeximal-string
 */
- (NSString *)nsStringFromNSData:(NSData *)data {
    const unsigned char *dataBuffer = (const unsigned char *)[data bytes];
    if (!dataBuffer)
        return nil;

    NSUInteger dataLength  = [data length];
    NSMutableString *hexString  = [NSMutableString stringWithCapacity:(dataLength * 2)];

    for (int i = 0; i < dataLength; ++i) {
        [hexString appendString:[NSString stringWithFormat:@"%02lx", (unsigned long)dataBuffer[i]]];
    }
    return [NSString stringWithString:hexString];
}

@end

File: Vialer/APNS/APNSHandler.swift
//
//  APNSHandler.swift
//  Vialer
//
//  Created by Chris Kontos on 05/06/2019.
//  Copyright © 2019 VoIPGRID. All rights reserved.
//

import Foundation
import PushKit
import UIKit


@objc class APNSHandler: NSObject, PKPushRegistryDelegate {
    
    // MARK: - Properties
    @objc static var sharedAPNSHandler = APNSHandler()
    @objc var voipRegistry: PKPushRegistry = PKPushRegistry(queue: nil)
    @objc var middleware: Middleware = Middleware()
        
    // MARK: - Lifecycle
    @objc private override init(){}
    
    @objc class func setSharedHandler(_ sharedHandler: APNSHandler) {
        if sharedAPNSHandler != sharedHandler {
            sharedAPNSHandler = sharedHandler
        }
    }
    let appDelegate = UIApplication.shared.delegate as! AppDelegate

    
    // MARK: - Actions
    @objc func registerForVoIPNotifications() {
        // Only register once, if delegate is set, registration has been done before:
        if voipRegistry.delegate == nil {
            voipRegistry.delegate = self
            
            VialerLogVerbose("Initiating VoIP push registration")
            voipRegistry.desiredPushTypes = Set([.voIP])
        }
    }
    
    @objc class func storedAPNSToken() -> String? {
        let sharedHandler: APNSHandler? = self.sharedAPNSHandler
        let token: Data? = sharedHandler?.voipRegistry.pushToken(for: .voIP)
        return sharedHandler?.nsString(fromNSData: token) as String?
    }
    
    // MARK: - PKPushRegistry management
    @objc func pushRegistry(_ registry: PKPushRegistry, didInvalidatePushTokenFor type: PKPushType) {
        VialerLogWarning("APNS Token became invalid")
    }
    
    func pushRegistry(_ registry: PKPushRegistry, didReceiveIncomingPushWith payload: PKPushPayload, for type: PKPushType, completion: @escaping () -> Void) {
        guard type == .voIP else { return }
        
        // Extract the call information from the push notification payload.
        if let phoneNumberString = payload.dictionaryPayload[PushedCall.MiddlewareAPNSPayloadKeyPhonenumber] as? String,
            let uuidString = payload.dictionaryPayload[PushedCall.MiddlewareAPNSPayloadKeyUniqueKey] as? String {
            // The uuid string in the payload is missing hyphens so fix that.
            let callUUID = NSUUID.uuidFixer(uuidString: uuidString)! as UUID
            
            // Configure the call information data structures.
            let callUpdate = CXCallUpdate()
            let phoneNumberHandle = CXHandle(type: .phoneNumber, value: phoneNumberString)
            callUpdate.remoteHandle = phoneNumberHandle
            callUpdate.localizedCallerName = phoneNumberString
            if let callerId = payload.dictionaryPayload[PushedCall.MiddlewareAPNSPayloadKeyCallerId] as? String {
                callUpdate.localizedCallerName = callerId
            }
        
            
            APNSCallHandler(payload: payload).handle(completion: completion, uuid: callUUID, number: phoneNumberString, update: callUpdate)
        }
    }
    
    @objc func pushRegistry(_ registry: PKPushRegistry, didUpdate credentials: PKPushCredentials, for type: PKPushType) {
        VialerLogInfo("Type:\(type). APNS registration successful.")
        middleware.sentAPNSToken(nsString(fromNSData: credentials.token) as String? ?? "")
    }
    
    // MARK: - Token conversion
    /*
     * Returns hexadecimal string of NSData. Empty string if data is empty.
     * http://stackoverflow.com/questions/1305225/best-way-to-serialize-a-nsdata-into-an-hexadeximal-string
     */
    @objc func nsString(fromNSData data: Data?) -> NSString? {
        guard
            let data = data
            else {
                return ""
        }
        let token = data.map { String(format: "%02.2hhx", $0) }.joined()
        return NSString(utf8String: token)
    }
}

